src/mapgen/treegen.h

```diff
@@ -35,7 +35,14 @@ namespace treegen {
 		UNBALANCED_BRACKETS
 	};
 
-	struct TreeDef {
+	struct TreeDef : public NodeResolver {
+		void resolveNodeNames();
+
+		// Needed since m_ndef can't be accessed in read_tree_def.
+		void setNodeDefManager(const NodeDefManager* ndef) {
+			NodeResolver::m_ndef = ndef;
+		}
```

Comment:
you're using the noderesolver class wrong. admittedly there is no docs and I had to find out how it works myself.

1. do not set `m_ndef`
2. call `pendNodeResolve` after populating the lists
3. `resolveNodeNames` must be a virtual function (it is defined in `NodeResolver`)
4. do not manually call this function
5. you do not need an extra code path to handle the non-registration case. NodeResolver works transparently at runtime.

Reply:
Thank you for the explanation, I didn't even know that `pendNodeResolve` existed. This makes total sense.

I fixed it. Now the node resolve things are only in `read_tree_def` and in `TreeDef.resolveNodeNames()`, such that only the `NodeDefManager` has to handle it.